Feature #189: Cookie banner

Agupane opened the pull request:
Closes #189 

# Description
- Implements cookie banner 
- Also, we were having problems with flow's warning and error, so we updated flow-types dependencys

## Looks like this now:
![image](https://user-images.githubusercontent.com/21086218/69543440-609ce100-0f6c-11ea-940c-4c3d2c38534e.png)
![image](https://user-images.githubusercontent.com/21086218/69543448-6397d180-0f6c-11ea-816d-874290d5f6c1.png)

gnosis-info commented:
Travis automatic deployment:
 https://pr296--safereact.review.gnosisdev.com 
 
 Storybook book automatic deployment: 
 https://pr296--safereactstorybook.review.gnosisdev.com

lukasschor commented:
Looks good.

1) Pressing on "Accept cookies" would be counted as a consent to "necessary" AND "analytics", even if "analytics" is not selected? Only closing the banner through the "x" would limit the consent to "necessary", right?
2) Could we have the checkmark for "necessary" locked in, as shown in the designs? So it's not actually clickable. Currently it's possible to de-select necessary and close the banner, which should not be possible.
3) I'll send you the link to the Cookie Policy as soon as possible (waiting for our DPO to approve it)

Agupane commented:
Hi @lukasschor !
1) No, if the user did not selected analytics, only necessary would be counted as a consent. If you close it through "X", yes, it will count only necessary as consent, event if the analytics were selected.
2) Of course, will do it
3) Sounds good thanks!

mikheevm commented:
>Also, we were having problems with flow's warning and error, so we updated flow-types dependencys 
How did you do that? I thought it was not fixable 😄

mikheevm commented on src/index.js:
```diff
@@ -8,15 +8,22 @@ import { store } from '~/store'
-ReactDOM.render(<Root />, document.getElementById('root'))
+const root = document.getElementById('root')
+
+if (root !== null) {
```
why do you need this check?

Agupane replied:
f we don’t do this, flow’s server will complain.
In this case root will never be null, but Flow doesn’t seem to know.

Agupane commented:
Hi @mikheevm !.

For updating the flow dependencies we did this:
```
- Deleted flow-typed folder
- flow-typed install
- Updated / fixed flowconfig file
- Updated / fixed eslintignore file
```

mikheevm commented:
> * #272

Sorry, I don't understand why we'd need it. Could you explain please? I thought we'd just load GA/Intercom script inside the cookiebanner component and then we can check localstorage in index.js

Agupane commented:
I prefer to not load those scripts inside the cookie banner to avoid mixing things, I would prefer to do it that separately. And do things like [this PR](https://github.com/gnosis/safe-react/pull/299). If you still don't like the idea I can change it

gabitoesmiapodo commented:
@Agupane @lukasschor 

The checkbox's `disabled` status should be fixed now.

gnosis-info commented:
Travis automatic deployment:
 https://pr296--safereact.review.gnosisdev.com 
 
 Storybook book automatic deployment: 
 https://pr296--safereactstorybook.review.gnosisdev.com

mikheevm commented:
> I prefer to not load those scripts inside the cookie banner to avoid mixing things, I would prefer to do it that separately. And do things like [this PR](https://github.com/gnosis/safe-react/pull/299). If you still don't like the idea I can change it

Why not? The entire point of redux is to put state there if we have to share it between few places in our application. In this case, state is consumed only in CookieBanner component.

Agupane commented:
It's not, it's consumed also in the google analytics component and will be on the intercom component. Anyways, I'm going to remove that from the store

gnosis-info commented:
Travis automatic deployment:
 https://pr296--safereact.review.gnosisdev.com 
 
 Storybook book automatic deployment: 
 https://pr296--safereactstorybook.review.gnosisdev.com

lukasschor commented:
> Hi @lukasschor !
> 
> 1. No, if the user did not selected analytics, only necessary would be counted as a consent. If you close it through "X", yes, it will count only necessary as consent, event if the analytics were selected.
> 2. Of course, will do it
> 3. Sounds good thanks!

Regarding 1: Clicking the "accept cookies" should also count as consent to "Analytics". That's also what we state in the text above: " By clicking “Accept Cookies,” you agree to the storing of cookies on your device to enhance site navigation and **analyze site usage.**

Agupane commented:
So, for clarifying, it will be like this:
```
1- Necessary checked + Analytics unchecked + Accept Cookies -> necessary && analytics enabled
2- Necessary checked + Analytics unchecked + Close Button: -> necessary enabled && analytics disabled
3- Necessary checked + Analytics Checked + Accept Cookies: -> necessary && analytics enabled
4- Necessary checked + Analytics Checked + Close Button: -> necessary && analytics enabled
```

gnosis-info commented:
Travis automatic deployment:
 https://pr296--safereact.review.gnosisdev.com 
 
 Storybook book automatic deployment: 
 https://pr296--safereactstorybook.review.gnosisdev.com

lukasschor commented:
As discussed in the call, let's do two changes:

1) To make it more clear to the user that clicking on accept would also be a constent to the analytics cookie, let's rename the button to "Accept all"

2) To make clear that analytics are required to provide customer support through intercom, let's change the last sentence in the explainer to 

> By clicking "Accept all", you agree to the storing of cookies on your device to enhance site navigation, analyze site usage and provide customer support.

Agupane commented:
I'd updated the PR with the last changes. It looks like this now:

![image](https://user-images.githubusercontent.com/21086218/69753277-f766ca80-1131-11ea-9fec-662bfdeaa975.png)

gnosis-info commented:
Travis automatic deployment:
 https://pr296--safereact.review.gnosisdev.com 
 
 Storybook book automatic deployment: 
 https://pr296--safereactstorybook.review.gnosisdev.com

mikheevm commented on package.json:
```diff
@@ -52,6 +52,7 @@
     "material-ui-search-bar": "^1.0.0-beta.13",
     "notistack": "https://github.com/gnosis/notistack.git#v0.9.4",
     "optimize-css-assets-webpack-plugin": "5.0.3",
+    "polish": "^0.2.3",
```
what's this lib used for? Couldn't find in the code (I could find `polished`)

gnosis-info commented:
Travis automatic deployment:
 https://pr296--safereact.review.gnosisdev.com

mikheevm commented on src/components/CookiesBanner/index.jsx:
```diff
@@ -0,0 +1,165 @@
+      acceptedNecessary: true,
+      acceptedAnalytics: true,
+    }
+    await saveToStorage(COOKIES_KEY, newState)
```
as mentioned in the call we had, the consent should be stored as a cookie (you can use `js-cookie` package for this) because we have to renew it (IIRC its 365 days for positive and 7 days for negative, but better to double check with legal)
cc @lukasschor

Agupane replied:
done

gnosis-info commented:
Travis automatic deployment:
 https://pr296--safereact.review.gnosisdev.com

mikheevm commented on src/theme/mui.js:
```diff
@@ -308,17 +310,31 @@ export default createMuiTheme({
+    MuiCheckbox: {
+      colorSecondary: {
+        '&$disabled': {
+          color: rgba(secondary, 0.5),
```
rgba is undefined, commit https://github.com/gnosis/safe-react/pull/296/commits/fe422cba9f053a41b498df209ce19e46c69d8229 removed it

mikheevm replied:
Package `polish` was removed because it was incorrect, instead we should use `polished`

mikheevm replied:
Also please update the flow types after removing `polish` 😄

gabitoesmiapodo replied:
This is weird, I don't know why there was a `polish` there (I suppose it was a typo). But Flow should've warned us (I guess) about the wrong import, and it didn't...

Maybe `flow-typed` overlooked it.

Anyways: I restored the import in `src/theme/mui.js`, added the correct library `"polished": "^3.4.2",`, and updated `flow-typed` files.

Thanks for your input.

gnosis-info commented:
Travis automatic deployment:
 https://pr296--safereact.review.gnosisdev.com

gnosis-info commented:
Travis automatic deployment:
 https://pr296--safereact.review.gnosisdev.com

lukasschor commented:
Yes, cookie renewal should be:
365 days for positive
7 days for negative

gnosis-info commented:
Travis automatic deployment:
 https://pr296--safereact.review.gnosisdev.com

mikheevm commented:
Sorry, we forgot about one more requirement. So by law the user should be able to change his consent any time. Because of this, we should have a button inside the interface which would open the cookie banner again. Let's do it in the sidebar at the bottom where we have all the legal links with a label "Cookies settings"

gnosis-info commented:
Travis automatic deployment:
 https://pr296--safereact.review.gnosisdev.com

Agupane commented:
Done, please check it now @mikheevm

gnosis-info commented:
Travis automatic deployment:
 https://pr296--safereact.review.gnosisdev.com

lukasschor commented:
Works well. However, I think it would be better UX if clicking on "Cookies" would also close the sidebar.

mikheevm commented on src/logic/cookies/store/reducer/cookies.js:
```diff
@@ -0,0 +1,24 @@
+    [OPEN_COOKIE_BANNER]: (state: State, action: ActionType<Function>): State => {
+      const { cookieBannerOpen } = action.payload
+
+      const newState = state.withMutations((map) => {
```
there's no real need in `.withMutations` because it performs only a single change of a state. `.withMutations` should be used when you want to perform many changes on the same immutable.js object for performacne reasons:
https://stackoverflow.com/questions/28510753/when-should-i-use-withmutations-on-a-map-in-immutable-js/29381227

gnosis-info commented:
Travis automatic deployment:
 https://pr296--safereact.review.gnosisdev.com

Agupane commented:
It does that now @lukasschor 

> Works well. However, I think it would be better UX if clicking on "Cookies" would also close the sidebar.

It does that now please check it :)

gnosis-info commented:
Travis automatic deployment:
 https://pr296--safereact.review.gnosisdev.com
